_posts/2016-4-20-The Fashion Logging System -- ELK - Part 1.md

```diff
@@ -0,0 +1,161 @@
+```
+curl -XPUT 'localhost:9200/<index: restaurant>
+```
+那講完怎麼放資料之後，就要來看怎麼搜尋資料啦，就簡單的就是用 REST API 來 GET 資料
+```
+curl 'localhost:9200/<index: restaurant>/<type: customer>/xxx' -d '
+```
+這樣就會獲得 restaurant 下的 customer 的 ID 為 xxx 的 document，不過上面有講了 ES 搜尋功能如此強大，當然不是區區 REST API 可以發揮的，而是要使用 DSL 的語法
+```
+curl -XPOST 'localhost:9200/<index: restaurant>/_search' -d '
+{
+  "query": { "match": { "name": "John" } }
+}'
+```
+這樣就會回傳 restaurant 這個 index 下所有符合 name 是 John 的 document，而詳細的 DSL 用法可以參考 [DSL Document]  
+Elasticsearch 的介紹大概到這邊，想暸解更多可以直接到 [Elastic 的官網上看 Document]  
+
+#### Kibana
+Kibana 簡單來說就是讓你更容易讀取 ES 資料的圖像化工具，將複雜的 DSL 包裝成可以直接用介面選擇的方式就完成 query 同時用 query 的結果畫出圖表呈現給別人看，可以說是沒有 Kibana 的話，ES 就像是失去雙手一樣不方便，不過當然你可以選擇自己做或用坊間的 Library 去對 ES 做讀取，其實不同語言都有一些 implement 甚至有語言有做 stream 的功能，這方面就讓大家自行摸索了，這邊主要介紹 Kibana 的結構，主要分成三層，分別是：discover, visualize, dashboard  
+- discover  
+discover 其實就是 search 的語法，例如 name = John(當然也可以使用複雜的 DSL Filter)，而 discover 就是負責把這些語法存起來，每次使用不同的 discover，在 discover 的頁面上就會有不同的 raw document 顯示在上面，而這些存起來的 discover 也是丟給 visualize 用的東西
+- visualize  
+visualize 主要的功能就是依據 discover filter 出來的資料來畫出圖表，你可以先選定要畫的圖表類型，接著選擇要使用的 discover，最後再根據你選的圖表類型去選擇 x 軸要用 discover filter 出來的 documents 的哪個 field ，以及該 field 要經過什麼運算才顯示在 x 軸，這樣就可以完成一張圖表，並且你可以將這張圖表儲存起來，要特別注意的是，圖表儲存的其實是 schema 而已，也就是使用哪個 discover 以及哪個 field 這樣的資訊，所以你每次看該圖表的時候都有可能隨著資料變多而改變圖表的樣貌
+- dashboard  
+最後是 dashboard，dashboard 的功能就是將 visualize 畫出來的圖表組合起來，一起顯示在同一個頁面，你一樣可以把做好的 dashboard 儲存起來，方便之後直接 load 出來使用  
+
+以上大概就是 Kibana 的大致介紹，有興趣的人一樣可以到 [Elastic 的 Kibana 網頁上觀看 Document]  
+
+#### Logstash
+既然介紹了怎麼用 Kibana 從 ES 裡讀取資料，那總是要想個辦法把資料丟到 ES 裡啊！當然我們可以直接使用 ES 提供的 REST API 去 POST Data，不過往往我們的 Log 其實都是非結構化的資料，直接丟進去的話即使用 Kibana 也畫不出什麼東西，所以勢必要有一個服務可以幫我們做 formatter 的工作，同時當我們的 Log 來源有好多個，要送去的 ES 也不同，甚至是有其它儲存 log 的地方的時候，統一由一個服務來做蒐集、format、發送的工作，會更容易管理，而這就是 Logstash 的場子了！  
+Logstash 基本上就像上面說的，負責蒐集之後做一些處理，接著把資料送到我們想要的地方，而 Logstash 的設定方式也很單純，就是安裝好之後給它一個 .conf 檔之後 start 就可以了，.conf 檔的結構分成三個部分: input、filter、output，而資料處理的流程則是 inputs → filters → outputs  
+
+- input  
+input 就是負責蒐集資料的 section，Logstash 的 input 方式提供了許多種，你可以讓它監測 file，當 file 內容有變更的時候它就會將變更的內容抓進來，你可以指定 HTTP，設定好 port 之後就可以直接用 HTTP 將資料送進 Logstash，甚至是你可以讓 Logstash 去抓 twitter 裡的訊息，範例如下
+```
+# logstash.conf
+input {
+  file {
+    path => "/tmp/access_log"
+  }
+  http {
+  	port => "8000"
+  }
+}
+```
+這樣 Logstash 就可以從 /tmp/access_log 的檔案裡讀取內容，同時從 HTTP 的 8000 port 讀取資料  
+更多 input plugin 請參考 [Logstash input plugin說明]
+
+- filter  
+把資料抓進來後，接著就是要對資料做處理，而 filter 就是負責這部分的，直接舉例說明，假設今天從 input 那邊吃進來了 Apache 的 Log，它一定長得醜醜一大串文字，像是：
+```
+127.0.0.1 - - [11/Dec/2013:00:01:45 -0800] "GET /xampp/status.php HTTP/1.1" 200 3891 "http://cadenza/xampp/navi.php" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.9; rv:25.0) Gecko/20100101 Firefox/25.0"
+```
+我們需要把它整理成一份 JSON 的格式的 document，像是 {"http_method":"GET"}，這樣子才比較容易做搜尋，而 Logstash 也提供的 filter 的 plugin叫過 grok，grok 可以自動幫你把 Apache 格式的 raw data 轉成結構化的資料格式
```

Comment:
Wrap `{"http_method":"GET"}` with backtick ( **`** ) for monospace font. (and also other similar cases)

```markdown
我們需要把它整理成一份 JSON 的格式的 document，像是 `{"http_method":"GET"}`，這樣子
```